mintpy/objects/cluster.py

```diff
@@ -1,102 +1,276 @@
-        msg += '\nsupported cluster types: {}'.format(cluster_list)
-        raise ValueError(msg)
-    print("dask cluster type: {}".format(cluster_type))
-
-    # import related dask module only
-    # because job_queue is not available on macports, which make sense
-    if cluster_type == 'local':
-        try:
-            from dask.distributed import LocalCluster
-        except ImportError:
-            raise ImportError('Cannot import dask.distributed.LocalCluster!')
-    else:
-        try:
-            from dask_jobqueue import LSFCluster, PBSCluster, SLURMCluster
-        except ImportError:
-            raise ImportError('Cannot import dask_jobqueue!')
-
-    # for local cluster, NO need to do the extra configuration
-    if cluster_type == 'local':
-        return LocalCluster()
-
-    # check input config name
-    if 'config_name' in kwargs.keys():
-        kwargs['config_name'] = format_config_name(kwargs['config_name'], cluster_type)
-    print("Dask config name: {}".format(kwargs['config_name']))
-
-    # check walltime format for each cluster type
-    if 'walltime' in kwargs.keys():
-        kwargs['walltime'] = format_walltime(kwargs["walltime"], cluster_type)
-    print('Dask worker walltime: {}'.format(kwargs['walltime']))
-
-    # initiate cluster object
-    if cluster_type == 'lsf':
-        cluster = LSFCluster(**kwargs)
-    elif cluster_type == 'pbs':
-        cluster = PBSCluster(**kwargs)
-    elif cluster_type == 'slurm':
-        cluster = SLURMCluster(**kwargs)
-
-    # Print and write job command file for HPC cluster types
-    print("JOB COMMAND CALLED FROM PYTHON:\n\n", cluster.job_script())
-    with open('dask_command_run_from_python.txt', 'w') as f:
-        f.write(cluster.job_script() + '\n')
-
-    return cluster
-
-
-def format_config_name(config_name, cluster_type):
-    # due to the pre-set in mintpy.yaml, default config_name is the same as cluster_type
-    if not config_name:
-        config_name = cluster_type
-
-    # check if config_name exists in ~/.config/dask/*.yaml
-    config_names = list(dask.config.get('jobqueue').keys())
-    if config_name not in config_names:
-        msg = 'Dask configuration "{}" was not found in ~/.config/dask/*.yaml'.format(config_name)
-        msg += '\nFall back to default config name: "{}"'.format(cluster_type)
-        print(msg)
-        config_name = cluster_type
-
-    return config_name
-
-
-def format_walltime(walltime, cluster_type):
-    """format the walltime str for different clusters
-    HH:MM:SS - pbs / slurm
-    HH:MM    - lsf
-    """
-    num_digit = len(walltime)
-    if cluster_type in ['pbs', 'slurm'] and num_digit != 8:
-        if num_digit == 5:
-            walltime += ":00"
+import glob
+import os
+import shutil
+import time
+import dask
+from dask.distributed import LocalCluster, Client, as_completed
+
+
+class DaskCluster:
+
+    def __init__(self, cluster_type, num_workers, write_job_script=True, **kwargs):
```

Comment:
Is `write_job_script` for the HPC clusters? I don't see this argument is assigned in ifgram_inversion.py, is this expected?

Reply:
I don’t understand the question.

Reply:
If this argument does not change, maybe we don't need it here.

Reply:
What argument are we referring to?

Reply:
`write_job_script `. Sorry to not specify this earlier!

Reply:
Honestly, that file gets written more for debugging purposes than anything, so once we have dask stable and running on most systems it can get turned off.

Reply:
Having debug code embeded is totally fine, as long as it's clearly commented and turned off by default. One example is [here](https://github.com/insarlab/MintPy/blob/master/mintpy/objects/resample.py#L474). Shall we do that?

Reply:
Yes. Lets do something like that.